content/6_Advanced/Offline_Del.mdx

```diff
@@ -1,100 +1,285 @@
+	vector<pair<ll &, ll>> history;
 
-<Warning title="Watch Out!">
+  public:
+	DSU(int n) {
+		p.resize(n);
+		sz.resize(n, 1);
+		sum.resize(n);
+		iota(p.begin(), p.end(), 0);
+	}
 
-Because path compression is amortized, it does not guarauntee
-$\mathcal{O}(N \log^2 N)$ runtime. You _must_ use merging by rank.
+	void init_sum(vector<ll> &a) {
+		for (int i = 0; i < (int)a.size(); i++) { sum[i] = a[i]; }
+	}
 
-</Warning>
+	int get(int x) { return (p[x] == x) ? x : get(p[x]); }
 
-### Implementation
+	ll get_sum(int x) { return sum[get(x)]; }
 
-<LanguageSection>
+	void unite(int a, int b) {
+		a = get(a);
+		b = get(b);
+		if (a == b) return;
+		if (sz[a] < sz[b]) { swap(a, b); }
 
-<CPPSection>
+		// add to history
+		history.push_back({p[b], p[b]});
+		history.push_back({sz[a], sz[a]});
+		history.push_back({sum[a], sum[a]});
 
-<!-- verified at https://codeforces.com/contest/1217/submission/62335472 -->
+		p[b] = a;
+		sz[a] += sz[b];
+		sum[a] += sum[b];
+	}
 
-```cpp
-int p[MN], r[MN];
-int *t[MN * 40], v[MN * 40], X;
-int setv(int *a, int b) {
-	if (*a != b) t[X] = a, v[X] = *a, *a = b, ++X;
-	return b;
+	void add(int x, ll v) {
+		x = get(x);
+		history.push_back({sum[x], sum[x]});
+		sum[x] += v;
+	}
+
+	int snapshot() { return history.size(); }
+
+	void rollback(int until) {
+		while (snapshot() > until) {
+			history.back().first = history.back().second;
+			history.pop_back();
+		}
+	}
+};
+// EndCodeSnip
+
+const int MAXN = 3e5 + 1;
+
+DSU dsu(MAXN);
+
+struct Query {
+	int t, u, v, x;
+};
+
+vector<Query> tree[MAXN * 4];
+
+void update(Query &q, int v, int l, int r, int L, int R) {
+	if (l > R || r < L) return;
+	if (l <= L && r >= R) {
+		tree[v].push_back(q);
+		return;
+	}
+	int m = (L + R) / 2;
+	update(q, v * 2, l, r, L, m);
+	update(q, v * 2 + 1, l, r, m + 1, R);
 }
-void rollback(int x) {
-	for (; X > x;) --X, *t[X] = v[X];
+
+void dfs(int v, int l, int r, vector<ll> &ans) {
+	int snapshot = dsu.snapshot();
+	// perform all available operations upon entering
+	for (Query &q : tree[v]) {
+		if (q.t == 1) { dsu.unite(q.u, q.v); }
+		if (q.t == 2) { dsu.add(q.v, q.x); }
+	}
+	if (l == r) {
+		// answer type 3 query if we have one
+		for (Query &q : tree[v]) {
+			if (q.t == 3) { ans[l] = dsu.get_sum(q.v); }
+		}
+	} else {
+		// go deeper into the tree
+		int m = (l + r) / 2;
+		dfs(2 * v, l, m, ans);
+		dfs(2 * v + 1, m + 1, r, ans);
+	}
+	// undo operations upon exiting
+	dsu.rollback(snapshot);
 }
-int find(int n) { return p[n] ? find(p[n]) : n; }
-bool merge(int a, int b) {
-	a = find(a), b = find(b);
-	if (a == b) return 0;
-	if (r[b] > r[a]) std::swap(a, b);
-	if (r[a] == r[b]) setv(r + a, r[a] + 1);
-	return setv(p + b, a), 1;
+
+int main() {
+	cin.tie(0)->sync_with_stdio(0);
+	int n, q;
+	cin >> n >> q;
+	vector<ll> a(n);
+	for (int i = 0; i < n; i++) { cin >> a[i]; }
+	dsu.init_sum(a);
+
+	map<pair<int, int>, int> index_added;
+	for (int i = 0; i < q; i++) {
+		int t;
+		cin >> t;
+		if (t == 0) {
```

Comment:
switches are very cool

Reply:
NO

Reply:
did you have a traumatic childhood incident with them what

Reply:
no one even knows what they are

Reply:
that's just false what